strawberry/utils/inspect.py

```diff
@@ -14,3 +15,88 @@ def get_func_args(func: Callable[[Any], Any]):
         for arg_name, param in sig.parameters.items()
         if param.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD
     ]
+
+
+@overload
+def get_specialized_type_var_map(
+    cls: type,
+    *,
+    include_type_vars: Literal[True],
+) -> Optional[Dict[TypeVar, Union[TypeVar, type]]]:
+    ...
+
+
+@overload
+def get_specialized_type_var_map(
+    cls: type,
+    *,
+    # NOTE: Union[bool, Literal[False]] is to make both mypy/pyright happy
+    include_type_vars: Union[bool, Literal[False]] = ...,
```

Comment:
what error where they throwing? :O

Reply:
If I just use `Literal[False]`, the tests will have some pyright errors because I have a parametrize value of type `bool` and it says that `bool` cannot be assigned to `Literal[True] | `Literal[False]`

If I try to make pyright happy and define it like this:

```python
@overload
def get_specialized_type_var_map(
    cls: type,
    *,
    include_type_vars: Literal[True],
) -> Optional[Dict[TypeVar, Union[TypeVar, type]]]:
    ...


@overload
def get_specialized_type_var_map(
    cls: type,
    *,
    include_type_vars: Literal[False],
) -> Optional[Dict[TypeVar, type]]:
    ...


@overload
def get_specialized_type_var_map(
    cls: type,
    *,
    include_type_vars: bool = ...,
) -> Optional[Dict[TypeVar, Union[TypeVar, type]]]:
    ...
```

Mypy gives this warning:

```
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Invalid output: 
E   Actual:
E     ../../home/bellini/dev/strawberry/strawberry/utils/inspect:30: error: Overloaded function signatures 2 and 3 overlap with incompatible return types  [misc] (diff)
E     main:9: note: Revealed type is "Union[builtins.str, None]" (diff)
E   Expected:
E     main:9: note: Revealed type is "Union[builtins.str, None]" (diff)
E   Alignment of first line difference:
E     E: main:9: note: Revealed type is "Union[builtins.str, None]"...
E     A: ../../home/bellini/dev/strawberry/strawberry/utils/inspect:30: error: Ov...
E        ^
```

I just noticed that when I added the generic clause, I mistakenly removed one of the overloads. I'm looking at it right now to try to solve it properly because when adding the overload again, even though pyright was fine with it, mypy wasn't

Reply:
Ok, found the issue. I assumed `Dict[TypeVar, type]` and `Dict[TypeVar, Union[TypeVar, Type]]` would be compatible, but they are not. The solution was to make a union of them both as the return

```python
@overload
def get_specialized_type_var_map(
    cls: type,
    *,
    include_type_vars: Literal[True],
) -> Optional[Dict[TypeVar, Union[TypeVar, type]]]:
    ...


@overload
def get_specialized_type_var_map(
    cls: type,
    *,
    include_type_vars: Literal[False] = ...,
) -> Optional[Dict[TypeVar, type]]:
    ...


@overload
def get_specialized_type_var_map(
    cls: type,
    *,
    include_type_vars: bool,
) -> Optional[
    Union[Optional[Dict[TypeVar, type]], Dict[TypeVar, Union[TypeVar, type]]]
]:
    ...
```

I'll send the fix together with some other comments soon